keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlResourceRepository.kt

```diff
@@ -183,90 +191,83 @@ open class SqlResourceRepository(
-          .limit(1)
-          .fetchOne()
-          ?.let { (json) ->
-            objectMapper.readValue<ApplicationEvent>(json)
-          }
-
-        if (event.javaClass == previousEvent?.javaClass) return@transaction
-      }
+  // todo: add sql retries once we've rethought repository structure: https://github.com/spinnaker/keel/issues/740
+  override fun appendHistory(event: ResourceEvent) {
+    // for historical reasons, we use the resource UID (not the ID) as an identifier in resource events
+    val uid = jooq.select(RESOURCE.UID).from(RESOURCE).where(RESOURCE.ID.eq(event.uid)).fetchOne(RESOURCE.UID)
+    doAppendHistory(event, uid)
+  }
 
-      txn
-        .insertInto(EVENT)
-        .set(EVENT.SCOPE, Scope.APPLICATION.name)
-        .set(EVENT.UID, event.application)
-        .set(EVENT.TIMESTAMP, event.timestamp.atZone(clock.zone).toLocalDateTime())
-        .set(EVENT.JSON, objectMapper.writeValueAsString(event))
-        .execute()
-    }
+  override fun appendHistory(event: ApplicationEvent) {
+    doAppendHistory(event, event.application)
   }
 
-  // todo: add sql retries once we've rethought repository structure: https://github.com/spinnaker/keel/issues/740
-  override fun appendHistory(event: ResourceEvent) {
+  private fun doAppendHistory(event: PersistentEvent, uid: String) {
     jooq.transaction { config ->
       val txn = DSL.using(config)
 
       if (event.ignoreRepeatedInHistory) {
         val previousEvent = txn
           .select(EVENT.JSON)
-          .from(EVENT, RESOURCE)
-          .where(EVENT.SCOPE.eq(Scope.RESOURCE.name))
-          .and(RESOURCE.ID.eq(event.id))
-          .and(RESOURCE.UID.eq(EVENT.UID))
+          .from(EVENT)
+          // look for resource events that match the resource...
+          .where(EVENT.SCOPE.eq(Scope.RESOURCE.name)
+            .and(EVENT.UID.eq(uid))
+          )
+          // ...or application events that match the application as they apply to all resources
+          .or(EVENT.SCOPE.eq(Scope.APPLICATION.name)
+            .and(EVENT_JSON_APPLICATION.eq(event.application))
+          )
           .orderBy(EVENT.TIMESTAMP.desc())
           .limit(1)
           .fetchOne()
           ?.let { (json) ->
-            objectMapper.readValue<ResourceEvent>(json)
+            objectMapper.readValue<PersistentEvent>(json) as? ResourceHistoryEvent
           }
 
         if (event.javaClass == previousEvent?.javaClass) return@transaction
       }
 
       txn
         .insertInto(EVENT)
-        .set(EVENT.SCOPE, Scope.RESOURCE.name)
-        .set(EVENT.UID, select(RESOURCE.UID).from(RESOURCE).where(RESOURCE.ID.eq(event.id)))
+        .set(EVENT.SCOPE, event.scope.name)
+        .set(EVENT.UID, uid)
         .set(EVENT.TIMESTAMP, event.timestamp.atZone(clock.zone).toLocalDateTime())
         .set(EVENT.JSON, objectMapper.writeValueAsString(event))
         .execute()
     }
   }
 
   override fun delete(id: String) {
+    // TODO: these should be run inside a transaction
```

Comment:
why don't you just put them in a transaction?

Reply:
Trying to avoid too many things in one PR, especially those not directly tied to the issue being addressed, as one Professor Burns taught me. 😉